org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/inlayhint/LSPLineContentCodeMining.java

```diff
@@ -0,0 +1,117 @@
+/*******************************************************************************
+ * Copyright (c) 2020 Red Hat Inc. and others.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+package org.eclipse.lsp4e.operations.inlayhint;
+
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jface.text.BadLocationException;
+import org.eclipse.jface.text.IDocument;
+import org.eclipse.jface.text.ITextViewer;
+import org.eclipse.jface.text.codemining.LineContentCodeMining;
+import org.eclipse.lsp4e.LSPEclipseUtils;
+import org.eclipse.lsp4e.LanguageServersRegistry.LanguageServerDefinition;
+import org.eclipse.lsp4e.LanguageServiceAccessor;
+import org.eclipse.lsp4j.InlayHint;
+import org.eclipse.lsp4j.InlayHintLabelPart;
+import org.eclipse.lsp4j.InlayHintRegistrationOptions;
+import org.eclipse.lsp4j.Position;
+import org.eclipse.lsp4j.ServerCapabilities;
+import org.eclipse.lsp4j.jsonrpc.messages.Either;
+import org.eclipse.lsp4j.services.LanguageServer;
+import org.eclipse.swt.events.MouseEvent;
+
+public class LSPLineContentCodeMining extends LineContentCodeMining {
+
+	private InlayHint inlayHint;
+	private final LanguageServer languageServer;
+	private final LanguageServerDefinition languageServerDefinition;
+	private final @Nullable IDocument document;
+
+	public LSPLineContentCodeMining(InlayHint inlayHint, IDocument document, LanguageServer languageServer,
+			LanguageServerDefinition languageServerDefinition, InlayHintProvider provider) throws BadLocationException {
+		super(toPosition(inlayHint.getPosition(), document), provider);
+		this.inlayHint = inlayHint;
+		this.languageServer = languageServer;
+		this.languageServerDefinition = languageServerDefinition;
+		this.document = document;
+		setLabel(getInlayHintString(inlayHint));
+	}
+
+	protected static @Nullable String getInlayHintString(@NonNull InlayHint inlayHint) {
+		Either<String, List<InlayHintLabelPart>> labels = inlayHint.getLabel();
+		if (labels.isLeft()) {
```

Comment:
but about using `Either#map` instead of if/else?

Reply:
I didn't know this feature, thanks for this suggestion. Fixed.